Disabled state patterns Doc (#2083)
hashicorp/design-system: website/docs/patterns/disabled-patterns/partials/guidelines/guidelines.md

## Permissions 

> When a user does not have permissions, hide the related actions, navigation items and views.

Reviewer: ```suggestion
When a user does not have permissions, hide the related actions, navigation items, and views.
```

### Hide/show relevant UI

In this example, the user's role is **viewer**, and they cannot deploy or destroy a cluster. An explanation is unnecessary as users do not require additional functionality for their day-to-day tasks.

> ![Image showing two dropdown with lists open, one with only one menu item that is reader specific, while the other has more options that allow for further capabilities.](/assets/patterns/disabled-patterns/permissions-example.png)

Reviewer: ```suggestion
![Image showing two open dropdown lists with number of items shown based on user permissions, one with only one menu item, while the other has more options that allow for further capabilities.](/assets/patterns/disabled-patterns/permissions-example.png)
```

> Similarly, if a user has the permissions to update their software, show the relevant UI for them to do just that.

Reviewer 2: This looks like a new addition? I'm a bit unclear on what this means, are referring to a user being able to change their role?

Reviewer 3: Because updating was part of the old documentation, I wanted to add it into the permissions section.

Ultimately yes, if the user has the proper permissions to update what ever it is, we will show them the relevant message to do just that.

Reviewer 2: I guess what's unclear to me is specifically "update their software", that seems pretty vague. Is there a term or phrase that we can use instead that is more clear?

Reviewer 3: Hmmmm... not entirely sure. Maybe more so "if a user has permissions to update..." maybe thats enough? Removing the "your software?"

Reviewer 2: Yeah, I think removing software helps, maybe also changing to "upgrade" from "update"

Reviewer 3: Resolving, updating jargon.

### Provide guidance

When a user lands on a page they can't access, indicate why it's unavailable and provide guidance so the user can fix the error. Consider displaying a call to action so the user doesn't feel stuck.

![Showing a 403 error page with information about what is going on, with a way to go back and a way to log in with a different token](/assets/patterns/disabled-patterns/permissions-example-3.png)

## Conditional availability

> When other steps must be taken to enable a feature or function, provide context in place of that feature so users can enable this feature.

Reviewer 2: [suggestion] I think this could be streamlined a bit so you aren't using "feature" so many times.

```suggestion
When other steps must be taken to enable a feature or function, provide context in place of that feature so users can enable it.
```

!!! Dont

In this example, the button is disabled without context, leaving the user confused and unaware of the steps to enable the feature.

![Showing an interface with a disabled button that doesn't describe why that button is disabled. This is not a recommended experience.](/assets/patterns/disabled-patterns/conditional-availability-example-disabled.png)

Disabled elements lack interactive states, preventing users from hovering, focusing, or clicking. For users with assistive technology, e.g., a screen reader, disabled elements (including their text) are almost entirely ignored.

!!!

!!! Do

Instead, we recommended replacing the disabled element with contextual guidance to help users understand why this feature isn't available, and providing them with clear steps to enable it.

> ![Showing an interface with a compact alert explaining why the expected behavior isn't there with guidance.](/assets/patterns/disabled-patterns/conditional-availability-example-explanation.png)

Reviewer 2: Same question here about the use of the `highlight` alert, should this be `neutral` instead?

Reviewer 3: Ill update to neutral.

Reviewer 3: Updated.

If possible, provide CTAs directly to the place where users can take immediate action.

!!!

## Upgrading

> When upgradable features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity. However, be mindful to not overdue it and make the user feel overly advertised to.

Reviewer: ```suggestion
When upgradable features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity. However, be mindful not to overdo it and make the user feel overly advertised to.
```

> When upgradable features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity. Be mindful however to not overdue it and make the user feel overly advertised to.

Reviewer 2: [suggestion] This might read a bit better and is more in line with the writing style used elsewhere

```suggestion
When upgradable features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity. However, be mindful to not overdue it and make the user feel overly advertised to.
```

There are two ways to showcase an upgradable feature:

> - Space to explain

Reviewer: What does "Space to explain" mean? Is it some kind of upgrade notice or call out?

The image included is also unclear to me. (I couldn't tell at first what the image was trying to point out and then figured it must be the "Health" box with the reddish border but it still wasn't quite clear. A more rough hand drawn looking circle around it may be more obvious.)

Reviewer 3: I'll explain what this is in the below comment.

- Interrupting an action

### Space to explain

> In this example, "Terraform Health" is a an upgradable feature and is contextually relevant in the workspace. The UI has space to entice the user by explaining a bit of what this service can do for them.

Reviewer: Again, the reference to "space" seems too generic and is difficult to understand to me.

Reviewer 3: I can elaborate

Reviewer 3: Resolving. Elaborated.

![Screen shot of Terraform workspace highlighting Terraform Health on the side bar below information.](/assets/patterns/disabled-patterns/upgrade-inline-example.png)

### Interrupting an action

In this example, a user is about to perform an action and notices something new within the list of options, one with a "Beta" badge. 

> ![Showing a dropdown opened, with two list items. One for creating a static secret, while another using an auto-rotating secret with a beta badge next to it.](/assets/patterns/disabled-patterns/upgrade-modal-example-1.png)

Reviewer: ```suggestion
![Showing an open dropdown list, with two list items. One for creating a static secret, and another using an auto-rotating secret with a beta badge next to it.](/assets/patterns/disabled-patterns/upgrade-modal-example-1.png)
```

> ![Showing a dropdown opened, with two list items. One for creating a static secret, while another using an auto-rotating secret with a beta badge next to it.](/assets/patterns/disabled-patterns/upgrade-modal-example-1.png)

Reviewer 3: I'll just take out the beta tag in it and the descriptor too.

Enticed, they click it to use this feature, however, they are not enrolled in this tier and are interrupted by a modal explaining this feature a little more.

![A modal explaining beta to the user, with a CTA to learn more, or close the modal.](/assets/patterns/disabled-patterns/upgrade-modal-example-2.png)

> Since the action is the only explanation a user has at the moment, providing a modal as an interruption to the action allows a user to understand more without fully committing. Only when a user wants to "learn more," they can then be lead to a page with more details.

Reviewer 2: [suggestion] This might introduce a bit of brevity here

```suggestion
Interrupting the action with a modal helps a user understand more without fully committing, and if more details are necessary they can be directed to an informational page.
```
I'm not really sure what you're trying to say in the first part of this sentence ("Since the action is the only explanation..."), but this seems to read a bit more concise. WDYT?

Reviewer 3: Your suggestion makes sense I'm adding it.

## Service outage

In instances where outages may occur, provide a clear message to users on what they can do and add a link to [our status page](https://status.hashicorp.com/).

![Displaying Hashicorp's 500 internal error, providing a retry button and a link to the the status page.](/assets/patterns/disabled-patterns/service-outage-example.png)

## Incomplete flow

In a form, always enable the submit button. When buttons are disabled, users are often left confused and frustrated because they can't move forward. Instead, show users what errors occurred so they can fix them and proceed. 

!!! Do

> When a user clicks "Create cluster" but has failed to enter a value into a `required` input, an error will display, informing the user how to fix the issue.

Reviewer: Do we want to use more actionable language for "Do's"?

IE: instead of just providing an example situation, tell them expressly what to do.

"If a user clicks a form submit button but has failed to enter a value in a required field, display an error informing them how to fix the issue."

Reviewer 3: Makes sense, ill update.

Reviewer 3: Updated to what you specified. Resolving.

> ![Image showing an input field with an error explaining what was missed in a form context.](/assets/patterns/disabled-patterns/incomplete-flow-example.png)

Reviewer: If more actionable language is included, as in the example text I added above, then a caption could be included with the image explaining this example case.

!!!

Learn more about [form validation patterns](/patterns/form-patterns?tab=validation).

## Quota limitation

> When a quota has been reached, hide the action allowing them to create more objects. If a user can increase the quota, replace the original action with an action that allows them to do so. If the user cannot increase the quota, e.g., they don't have permission, direct them to a form.

Reviewer: ```suggestion
When the quota has been reached, hide the action allowing the user to create more objects. If they can opt to increase the quota, replace the original action with an action permitting them to do so. If the user cannot increase the quota, e.g., they don't have permission, direct them to a form.
```

> When a quota has been reached, hide the action allowing them to create more objects. If a user can increase the quota, replace the original action with an action that allows them to do so. If the user cannot increase the quota, e.g., they don't have permission, direct them to a form.

Reviewer 2: [question] when you say "direct them to a form" what do you mean here? Do you mean direct them to a form where they can change permissions?

Reviewer 3: This is coming straight from HCP's guidelines, and it was intentionally left open ended because it could be to change permissions, contact an admin, or really anything in between (could also be to input payment details). Initially I changed it to "contact an admin," but was reverted back because that wasn't in the initial guidelines.

Reviewer 2: Ok got it, thanks for the clarification there!

> ![Image showing a title with a button that allows a user to create a new service, with a table below it.](/assets/patterns/disabled-patterns/create-new-limitation-example.png)

Reviewer: ```suggestion
![Image showing a table of services with a button allowing the user to create a new service.](/assets/patterns/disabled-patterns/create-new-limitation-example.png)
```

Once the quota has been reached, the "Create" action is hidden, and an [`Alert`](https://helios.hashicorp.design/components/alert) is displayed, indicating why new objects can no longer be created.

> ![Image showing a title with an alert message below and a table below that.](/assets/patterns/disabled-patterns/create-new-limitation-reached-example.png)

Reviewer: ```suggestion
![Image showing a table of services with an alert message explaining that the services limitation has been reached and providing them with an upgrade link.](/assets/patterns/disabled-patterns/create-new-limitation-reached-example.png)
```

!!! Dont

In this example, the button is disabled without context, leaving the user confused and unaware of the steps to enable the feature.

![Showing an interface with a disabled button that doesn't describe why that button is disabled. This is not a recommended experience.](/assets/patterns/disabled-patterns/conditional-availability-example-disabled.png)

Disabled elements lack interactive states, preventing users from hovering, focusing, or clicking. For users with assistive technology, e.g., a screen reader, disabled elements (including their text) are almost entirely ignored.

!!!

!!! Do

Instead, we recommended replacing the disabled element with contextual guidance to help users understand why this feature isn't available, and providing them with clear steps to enable it.

![Showing an interface with a compact alert explaining why the expected behavior isn't there with guidance.](/assets/patterns/disabled-patterns/conditional-availability-example-explanation.png)

If possible, provide CTAs directly to the place where users can take immediate action.

!!!

## Upgrading

> When upgradable features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity. However, be mindful not to overdo it and make the user feel overly advertised to.

Reviewer: ```suggestion
When additional features are available if the user upgrades, highlighting a path to upgrade provides a graceful marketing opportunity. However, be mindful not to overdo it and make the user feel overly advertised to.
```

Reviewer 3: This is saying after upgrade. it should be before upgrade that highlighting a pathforward. I'm keeping the before for this.

There are two ways to showcase an upgradable feature:

> - Space to explain

Reviewer: I still think "Space to explain" is somewhat ambiguous and unclear.

How about something more along the lines of:
Feature highlight, Feature explanation, Feature spotlight
or similar?

Reviewer: Btw, the update to the related image makes it _much_ clearer. Thanks!

Reviewer 3: I like "feature spotlight" ill update!

> ### Permissions 

Reviewer 3: Will update.

> When a user does not have CRUD permissions, hide those related actions.

Author: I'd imagine no. I personally had to look it up when I was transferring the docs. Maybe adding a link for consumers who need a better understanding would be helpful. 

<img width="491" alt="Screenshot 2024-05-20 at 9 27 12 AM" src="https://github.com/hashicorp/design-system/assets/124841193/aea76f7a-8bbf-44b1-98f6-d9b346776d6d">

> When a user does not have CRUD permissions, hide those related actions.

Reviewer 3: Updating suggestion even more - thanks for the heads up:

```suggestion
When a user does not have CRUD (Create, Read, Update, Delete) permissions, hide the related actions.
```

> In this example, the user's role is **viewer** and they cannot **deploy** or **destroy** a cluster. An explanation is not needed as users understand that their account has limited permissions and that they require an additional user's help to gain these permissions (often someone with admin or similar roles).

Reviewer 2: [suggestion] flipped the order of these which _seems_ to read a bit better.

```suggestion
In this example, the user's role is **viewer** and they cannot **deploy** or **destroy** a cluster. An explanation is not needed as users understand that their account has limited permissions and that an additional user's help is required to obtain these permissions (often someone with admin or similar roles).
```

![Image showing two dropdown with lists open, one with only one menu item that is reader specific, while the other has more options that allow for further capabilities.](/assets/patterns/disabled-patterns/permissions-example.png)

> In scenarios where a user lands on a page they cannot access, provide a reason why with guidance so that users understand what steps they can take to rectify the error. So that users don't feel stuck, provide them a call to action to facilitate a path forward.

Reviewer 2: [suggestion] This adds a bit of brevity I think, but I'm not overly opinionated about it.

```suggestion
In scenarios where a user lands on a page they cannot access, provide a reason why with guidance and steps to rectify the error. So that users don't feel stuck, provide them a call to action to facilitate a path forward.
```

![Showing a 403 error page with information about what is going on, with a way to go back and a way to log in with a different token](/assets/patterns/disabled-patterns/permissions-example-3.png)

### Conditional availability

> When a feature or function is conditionally available depending on other steps that must be taken, provide context in place of that action so that users are able to take those steps to enable this feature.

Reviewer 2: This still reads very strange to me and is a bit jumbled. I _think_ I understand what you're trying to say, let me take a stab:

```suggestion
When a feature or function is unavailable due to unmet conditions, provide context in the form of text or an [Alert](/components/alert) guiding user through enabling it.
```

Could also alter the last part to be:

"...provide context or instructions guiding the user through enabling it." dropping the specific component/element usage.

!!! Dont

> In this example, the button is disabled and no context is provided, leaving the user confused and unaware of the steps required to enable the feature.

Reviewer 2: [suggestion] Maybe try this for brevity:

```suggestion
In this example, the button is disabled without context, leaving the user confused and unaware of the steps to enable the feature.
```

![Showing an interface with a disabled button that doesn't describe why that button is disabled. This is not a recommended experience.](/assets/patterns/disabled-patterns/conditional-availability-example-disabled.png)

> Disabled elements lack interactive states, preventing a user from hovering, focusing, or clicking them. For users with assistive technology (e.g., a screen reader), disabled elements are almost entirely ignored, including the text within.

Reviewer 2: I'm not opinionated on this, but for structure, does it make sense to move this out of the don't block?

Reviewer 3: Tried removing it out of and it looked weird. Opting to keep it in.

> Disabled elements lack interactive states, preventing a user from hovering, focusing, or clicking them. For users with assistive technology (e.g., a screen reader), disabled elements are almost entirely ignored, including the text within.

Reviewer 3: Maybe we removed "almost?"

Also good question for @MelSumner

!!!

!!! Do

Instead, we recommended replacing the disabled element with contextual guidance to help users to understand why this feature isn't available, and providing them with clear steps to enable it.

![Showing an interface with a compact alert explaining why the expected behavior isn't there with guidance.](/assets/patterns/disabled-patterns/conditional-availability-example-explanation.png)

If possible, provide CTAs directly to the place where users can take immediate action.

!!!

> ### Upgrading

Reviewer 3: What do you mean by available in this context, because if they don't have permissions to upgrade isn't that essentially say it is not available?

> ### Upgrading

Reviewer 3: Resolving because non-blocking.

> When premium features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity.

Reviewer 4: **question:** Where did the term "premium" come from? It's not in the original guidance, and I don't think I've ever heard of us offering "premium" features before.

Reviewer 3: The term was used specifically to indicate anything that requires additional payment. Is there a word you prefer to use? Maybe "When upgraded features..."?

Reviewer 3: I updated wording to say "upgraded feature"

> When premium features are available and a user has the necessary permissions, highlighting a path to upgrade provides a graceful marketing opportunity.

Reviewer 4: **question:** It feels as though we may have changed some of the original guidance. Here the guideline is saying that "highlighting a path to upgrade provides a graceful marketing opportunity." while the original guidance reads:

> In some cases, it may also be worth informing to users what they could do if they did have permission, but be mindful of not annoying users or making them feel advertised to. 

Why have we opted to change the language here?

Reviewer 3: Maybe I need to elaborate further here. I think there are two things we need to outline:

1. If the only way of communicating a feature is the action itself, then having a "midway point" like a modal upon click may communicate the limited access to this feature unless they upgrade.
2. If there is more space, then informing a user of what they could get if they upgraded would be the path here. An example of this is "Terraform Health" as [shown here](https://hashicorp.slack.com/archives/C79TH5JTF/p1716406282326699?thread_ts=1716405687.011639&cid=C79TH5JTF).

So maybe I should expand on this section futher.

Reviewer 3: I also noticed you mentioned that both "Updating" and "Feature not ready" are not in the docs, but I removed them because they would be repetitive of the "Permissions" section. Do you think I should re-add them even though they _may_ be redundant?

Reviewer 3: Resolving because non-blocking. also updated the language to "upgradable feature."

An example of this is Figma's "create branch" function that is only available at the enterprise tier. Clicking "create a branch" in Figma opens a marketing modal that entices users to upgrade to gain access to this feature.

> ![Screen shot of figma's modal showing information about how to upgrade to enterprise to gain access to creating branches.](/assets/patterns/disabled-patterns/upgrade-modal-example.png)

Reviewer 3: Done!

### Service outage

In instances where outages may occur, provide a clear message to users on what they can do and add a link to [our status page](https://status.hashicorp.com/).

An example of this is Instagram's login page. When a user has no connection to their servers, a message will appear on submit.

> ![Displaying Instagram's login page with an error below notifying the user they have no connection to instagram.](/assets/patterns/disabled-patterns/service-outage-example.png)

Reviewer 5: Not trying to be difficult here, but I don't recall us using external examples in our documentation (either good or bad); I would be skeptical about introducing such references, especially as we don't know much about their origins and reasoning.

Reviewer 4: This is a good point, @alex-ju! We've discussed this in Design Review, and @andgen404 is going to try to find some relevant internal examples to use instead.

Reviewer 3: Very good point, I'll swap this out and use a more HashiCorp example :)

Reviewer 3: Replaced with our 500 error example. Resolving.

### Incomplete flow

> Regardless of the progress of a form, always enable submission, even if there appear to be obvious steps missing. This allows a user to proceed in their workflow, even if it results in an error. When we disable the only way to move forward, users are often left confused and frustrated. Instead, allow users to understand what erorrs took place so that they can fix them and proceed in the form's completion.

Reviewer 3: That's fair, and yes you're right. I can try to update the language a bit to be more clear.

> In a form, regardless of whether all of the fields have been filled out, always enable the submit button. When we disable the only way to move forward, users are often left confused and frustrated. Instead, allow users to understand what erorrs took place so that they can fix them and proceed in the form's completion.

Reviewer 2: [suggestion] I think removing "we" here is a bit more in our writing style, and some other things that seem to clean this up a bit, though I think removing the "we" and fixing the typo are the most important.

```suggestion
In a form, always enable the submit button, regardless of whether all fields are filled out. Disabling the only path forward often leaves users confused and frustrated. Instead, communicate to users what errors occurred so they can fix them and complete the form.
```

> In this example, creating a cluster has required inputs. Clicking "create cluster" will display errors where applicable, allowing the user to understand where the problems are to remediate them.

Reviewer 2: [suggestion] This uses a bit more plain language

```suggestion
In this example, creating a cluster has required inputs. Clicking "create cluster" will display any errors, informing the user of errors and how to fix them.
```

Reviewer 2: Also, I'm not sure if this got lost somewhere (I don't see the comment anymore), but this paragraph references the entire form (click create cluster, other fields as requirements), but only shows one field. I thought I had left a previous note about considering adding more of the UI around this example.

Reviewer 3: Yeah, ill add more UI. Thanks for the suggestion :)

Reviewer 3: Updated UI and text to reflect the addition in the image.

> In this example, creating a cluster has required inputs. Clicking "create cluster" will display errors where applicable, allowing the user to understand where the problems are to remediate them.

Reviewer 3: Jory bought this up too. Will be adding the rest of the UI (or at least more).

> In this example, creating a cluster has required inputs. Clicking "create cluster" will display errors where applicable, allowing the user to understand where the problems are to remediate them.

Reviewer 3: Added more UI.

![Image showing an input field with an error explaining what was missed.](/assets/patterns/disabled-patterns/incomplete-flow-example.png)

### Quota limitation

> When a quota limit has been hit, hide the action and replace it with guidance and a call to action to increase the quota (if possible). If the user does not have permissions to upgrade, direct them to an admin instead.

Reviewer 4: **suggestion (non-blocking):** More verbose, but provides more clarity. 
```suggestion
When a quota has been reached, hide the action allowing them to create more objects. If a user can increase the quota, replace the original action with an action that allows them to do so. If the user cannot increase the quota, e.g., they don't have permission, direct them to an admin.
```

**question:** Should we really direct them to an admin or advocate for hiding the action altogether?

Reviewer 3: That's a good point, ill update to say "direct them to a form" instead of an admin.

Reviewer 3: Updated wording.

> In this example, there is a limit in creating a new "service." When the limit hasn’t been exceeded, the button is present and functions as normal.

Reviewer 4: **polish (non-blocking):**
```suggestion
When a quota hasn't been reached, the "Create" action is available and interactive. 
```

![Image showing a title with a button that allows a user to create a new service, with a table below it.](/assets/patterns/disabled-patterns/create-new-limitation-example.png)

> Once the user has reached the maximum number of services, the button is hidden, and a message is placed below indicating why they are no longer able to create a new service.

Reviewer 4: **polish (non-blocking):**
```suggestion
Once the quota has been reached, the "Create" action is hidden, and an [`Alert`](https://helios.hashicorp.design/components/alert) is displayed, indicating why new objects can no longer be created.
```

> Once the user has hit the max limitation of creating a new "service," the button is hidden, and a message is placed below indicating why they are no longer able to create a new service. 

Reviewer 2: ```suggestion
Once the user has reached the maximum number of services, the button is hidden, and a message is placed below indicating why they are no longer able to create a new service. 
```

> ![Image showing a title with an alert message below and a table below that.](/assets/patterns/disabled-patterns/create-new-limitation-reached-example.png)

Reviewer 3: Yeah, because this will never lead to an error, instead it's highlighting limits and capacities.

> ![Image showing a title with an alert message below and a table below that.](/assets/patterns/disabled-patterns/create-new-limitation-reached-example.png)

Reviewer 3: I'll update to neutral

> ![Image showing a title with an alert message below and a table below that.](/assets/patterns/disabled-patterns/create-new-limitation-reached-example.png)

Reviewer 3: Updated.
